File: android/src/main/java/so/prelude/reactnative/sdk/PreludeReactNativeSdkModule.kt
package so.prelude.reactnative.sdk

import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import so.prelude.android.sdk.Configuration
import so.prelude.android.sdk.Configuration.Companion.DEFAULT_REQUEST_TIMEOUT
import so.prelude.android.sdk.Endpoint
import so.prelude.android.sdk.Prelude
import java.net.URL

class PreludeReactNativeSdkModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("PreludeReactNativeSdk")

        AsyncFunction("dispatchSignals") Coroutine { sdkKey: String, endpointUrl: String?, timeoutMilliseconds: Long? ->
            dispatchSignals(endpointUrl, sdkKey, timeoutMilliseconds)
        }

        AsyncFunction("verifySilent") Coroutine { sdkKey: String, requestUrl: String ->
            verifySilent(
                sdkKey = sdkKey,
                requestUrl = requestUrl
            )
        }
    }

    private suspend fun dispatchSignals(
        endpointUrl: String?,
        sdkKey: String,
        timeoutMilliseconds: Long?
    ): String {
        val endpoint: Endpoint = endpointUrl?.let {
            Endpoint.Custom(it)
        } ?: Endpoint.Default

        val context = appContext.reactContext
            ?: throw IllegalStateException(
                "Invalid React Android Context. Cannot dispatch signals"
            )

        val config = Configuration(
            context.applicationContext,
            sdkKey,
            endpoint,
            requestTimeout = timeoutMilliseconds ?: DEFAULT_REQUEST_TIMEOUT
        )
        val prelude = Prelude(config)
                
        return prelude.dispatchSignals().getOrThrow()
    }

    private suspend fun verifySilent(
        sdkKey: String,
        requestUrl: String,
    ): String {
        val context = appContext.reactContext
            ?: throw IllegalStateException(
                "Invalid React Android Context. Cannot perform silent verification"
            )

        if (sdkKey.isBlank() || requestUrl.isBlank()) {
            throw IllegalArgumentException("SDK Key and Request URL must both be provided.")
        }

        val prelude = Prelude(context.applicationContext, sdkKey)

        return prelude.verifySilent(URL(requestUrl)).getOrThrow()
    }
}

File: packages/prelude/ios/PreludeReactNativeSdkModule.swift
import ExpoModulesCore

public class PreludeReactNativeSdkModule: Module {
    public func definition() -> ModuleDefinition {
        Name("PreludeReactNativeSdk")

        AsyncFunction("dispatchSignals") { (sdkKey: String, endpointUrl: String?, timeoutMilliseconds: Int64?, implementedFeaturesRawValue: Int64?, maxRetries: Int?) in
            let endpoint = endpointUrl != nil ? Endpoint.custom(endpointUrl!) : .default
            let timeout = timeoutMilliseconds != nil ? TimeInterval(timeoutMilliseconds!) / 1000 : 10.0

            let implementedFeatures: Features
            if let rawValue = implementedFeaturesRawValue {
                let uintRawValue = rawValue >= 0 ? rawValue : 0
                implementedFeatures = Features(rawValue: UInt64(uintRawValue))
            } else {
                implementedFeatures = []
            }
            let maxRetries = maxRetries ?? 3

            let configuration = Configuration(
                sdkKey: sdkKey,
                endpoint: endpoint,
                implementedFeatures: implementedFeatures,
                timeout: timeout,
                maxRetries: maxRetries
            )

            let prelude = Prelude(configuration)

            return try await prelude.dispatchSignals()
        }

        AsyncFunction("verifySilent") { (sdkKey: String, requestUrl: String) in
            if sdkKey.isEmpty || requestUrl.isEmpty {
                throw Exception(name: "IllegalArguments",
                                description: "SDK Key and Request URL must both be provided.",
                                code: "ILLEGAL_ARGUMENTS_EMPTY_FIELD")
            }

            let configuration = Configuration(
                sdkKey: sdkKey
            )

            let prelude = Prelude(configuration)

            guard let url = URL(string: requestUrl) else {
                throw Exception(name: "IllegalArguments",
                                description: "Request URL must be a valid URL.",
                                code: "ILLEGAL_ARGUMENTS_INVALID_URL")
            }

            return try await prelude.verifySilent(url: url)
        }
    }
}
